src/modules/audio/audio.c

```diff
@@ -440,4 +454,45 @@ struct SoundData* lovrAudioCapture(uint32_t frameCount, SoundData *soundData, ui
   }
 
   return soundData;
+}
+
+void lovrAudioGetDevices(AudioDevice **outDevices, size_t *outCount) {
+  if(state.deviceInfos) 
+    free(state.deviceInfos);
+  
+  ma_result gettingStatus = ma_context_get_devices(&state.context, NULL, NULL, NULL, NULL);
+  lovrAssert(gettingStatus == MA_SUCCESS, "Failed to enumerate audio devices: %d", gettingStatus);
+  *outCount = state.context.playbackDeviceInfoCount + state.context.captureDeviceInfoCount;
+  *outDevices = state.deviceInfos = calloc(*outCount, sizeof(AudioDevice));
+  for(int i = 0; i < *outCount; i++) {
+    ma_device_info *mainfo = &state.context.pDeviceInfos[i];
+    AudioDevice *lovrInfo = &state.deviceInfos[i];
+    lovrInfo->name = mainfo->name;
+    lovrInfo->type = i < state.context.playbackDeviceInfoCount ? AUDIO_PLAYBACK : AUDIO_CAPTURE;
+    lovrInfo->isDefault = mainfo->isDefault; // remove _private after bumping miniaudio
+    lovrInfo->identifier = &mainfo->id;
+    lovrInfo->minChannels = mainfo->minChannels;
+    lovrInfo->maxChannels = mainfo->maxChannels;
+  }
+}
+
+void lovrAudioUseDevice(AudioDeviceIdentifier identifier, int sampleRate, SampleFormat format) {
+  int deviceCount = state.context.playbackDeviceInfoCount + state.context.captureDeviceInfoCount;
+  for(int i = 0; i < deviceCount; i++) {
+    if (identifier == &state.context.pDeviceInfos[i].id) {
+      AudioType type = i < state.context.playbackDeviceInfoCount ? AUDIO_PLAYBACK : AUDIO_CAPTURE;
+      state.config[type].device = identifier;
+      if (sampleRate) state.config[type].sampleRate = sampleRate;
+      if (format != SAMPLE_INVALID) state.config[type].format = format;
```

Comment:
Hm, I disagree. If the code says to use 44.1khz at some point, my code is going to continue wanting to receive 44.1khz audio even if I later change device.

What this IS indicative of though is that perhaps useDevice() and setFormat() should be two separate calls, rather than one. Why does useDevice() set format anyway? What was I thinking?

Okay maybe I'll just make a setCaptureFormat(). That would also remove the confusion that makes it look like you can set playback format.

Reply:
I have now implemented setCaptureFormat instead.